sky/serve/serve_utils.py

```diff
@@ -0,0 +1,753 @@
+            # than the requested resources.
+            if requested_controller_resources.less_demanding_than(r):
+                controller_available = True
+                # Don't break here since we still want to find the max
+                # memory requirements.
+            # Remove the '+' in memory requirement.
+            max_memory_requirements = max(max_memory_requirements,
+                                          float(r.memory.strip('+')))
+    if controller_available:
+        # Determine max number of services on this controller.
+        max_services_num = int(max_memory_requirements /
+                               constants.SERVICES_MEMORY_USAGE_GB)
+        # Get current number of services on this controller.
+        services_num_on_controller = len(
+            global_user_state.get_services_from_controller_name(
+                controller_name))
+        # Only consider controllers that have available slots for services.
+        if services_num_on_controller < max_services_num:
+            return services_num_on_controller
+    return None
+
+
+def get_available_controller_name(
+        controller_resources: 'sky.Resources') -> Tuple[str, bool]:
+    """Get available controller name to use.
+
+    Only consider controllers that satisfy the requested controller resources,
+    and have available slots for services.
+    If multiple controllers are available, choose the one with most number of
+    services to decrease the number of controllers.
+
+    Args:
+        controller_resources: The resources requested for controller.
+
+    Returns:
+        A tuple of controller name and a boolean value indicating whether the
+        controller name is newly generated.
+    """
+    # Get all existing controllers.
+    existing_controllers = get_existing_controller_names()
+    available_controller_to_service_num = dict()
+    # Get a mapping from controller name to number of services on it.
+    for controller_name in existing_controllers:
+        services_num_on_controller = (
+            _get_service_num_on_controller_if_available(controller_name,
+                                                        controller_resources))
+        if services_num_on_controller is not None:
+            available_controller_to_service_num[controller_name] = (
+                services_num_on_controller)
+    if not available_controller_to_service_num:
+        new_controller_name = generate_controller_cluster_name(
+            existing_controllers)
+        # This check should always be true since we already checked the
+        # service name is valid in `sky.serve_up`.
+        clouds.Cloud.check_cluster_name_is_valid(new_controller_name)
+        return new_controller_name, True
+    # If multiple controllers are available, choose the one with most number of
+    # services.
+    return max(available_controller_to_service_num,
+               key=lambda k: available_controller_to_service_num[k]), False
```

Comment:
Yes, it will. This is because the behaviour of directly iterating on a dict is iterating its key set. Several tests:
```python
>>> d = {'a': 1, 'b': 2, 'c': 3}
>>> max(d, key=lambda k: d[k])
'c'
>>> for k in d:
...     print(k)
... 
a
b
c
```
But agree that it is hard to read... I'll change it to the following instead:
```python
max(available_controller_to_service_num.keys(),
    key=lambda k: available_controller_to_service_num[k])
```
(The reason not use `existing_controller` is there might be some controller not in `available_controller_to_service_num`